### PR TITLE
layer number bug fixes

### LINKED ISSUE #294: Check that the layer number is not 0
This parameter https://github.com/NVIDIA/TransformerEngine/blob/574f1b41e7473c3108b305b3ddef271392ddd9d5/transformer_engine/pytorch/attention.py#L719-L721

should indicate that the number must be 1-index based. An assertion could be added to enforce it.

Otherwise, if the user accidentally passes `0`, then https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention.py#L206 will raise a division by 0 error after this multiplication: https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention.py#L198

### PR DESCRIPTION
**Pull request by ksivaman**
Fixes #294, `UnfusedDotProductAttention` for `fp16`, and KV inference cache.

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
LGTM, I don't see any behavior change as long as a valid `layer_number` is provided.

That said, can you clarify the behavior of `layer_number`? It's still a bit opaque to me. From what I can see:
- It is used as a key in the KV inference cache:
https://github.com/NVIDIA/TransformerEngine/blob/706d7da3bd8bb3732188fd9c776511b836872758/transformer_engine/pytorch/attention.py#L1151-L1171
For this role, `layer_number` can be any hashable type as long as it is unique to each attention module.
- It is a scaling factor to prevent FP16 overflows when computing attention scores:
https://github.com/NVIDIA/TransformerEngine/blob/706d7da3bd8bb3732188fd9c776511b836872758/transformer_engine/pytorch/attention.py#L196-L198
I don't understand this though. Each value is the dot product of two `(hidden_dim/num_heads)` vectors, so I'd imagine the values would be independent of the number of layers, sequence length, batch size, etc.

**Comment by ksivaman** [comment]
@timmoon10 in the second scenario, `layer_number` is used as a scaling technique for better convergence with fp16 training.

**Comment by timmoon10** [comment]
Can you give some intuition or documentation? All layers should have QKV values around the same range due to layernorm, so the QK GEMM should also have a consistent range and not require scaling. Perhaps in bprop there is some growth or decay in the gradients, but why do we scale by `layer_number` instead of `1/layer_number` or `exp(+/- layer_number)`?
